arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt

```diff
@@ -0,0 +1,242 @@
+   * [fold] the [Effect] into an [Either]. Where the shifted value [R] is mapped to [Either.Left], and
+   * result value [A] is mapped to [Either.Right].
+   */
+  public suspend fun toEither(): Either<R, A> = fold({ Either.Left(it) }) { Either.Right(it) }
+
+  /**
+   * [fold] the [Effect] into an [Ior]. Where the shifted value [R] is mapped to [Ior.Left], and
+   * result value [A] is mapped to [Ior.Right].
+   */
+  public suspend fun toIor(): Ior<R, A> = fold({ Ior.Left(it) }) { Ior.Right(it) }
+
+  /**
+   * [fold] the [Effect] into an [Validated]. Where the shifted value [R] is mapped to
+   * [Validated.Invalid], and result value [A] is mapped to [Validated.Valid].
+   */
+  public suspend fun toValidated(): Validated<R, A> =
+    fold({ Validated.Invalid(it) }) { Validated.Valid(it) }
+
+  /**
+   * [fold] the [Effect] into an [Option]. Where the shifted value [R] is mapped to [Option] by the
+   * provided function [orElse], and result value [A] is mapped to [Some].
+   */
+  public suspend fun toOption(orElse: suspend (R) -> Option<A>): Option<A> = fold(orElse, ::Some)
+
+  /**
+   * [fold] the [Effect] into an [A?]. Where the shifted value [R] is mapped to
+   * [null], and result value [A].
+   */
+  public suspend fun orNull(): A? = fold({ null }, ::identity)
+
+  /** Runs the [Effect] and captures any [NonFatal] exception into [Result]. */
+  public suspend fun attempt(): Effect<R, Result<A>> = effect {
+    try {
+      Result.success(bind())
+    } catch (e: Throwable) {
+      Result.failure(e.nonFatalOrThrow())
+    }
+  }
+
+  public suspend fun handleError(recover: suspend (R) -> A): Effect<Nothing, A> = effect {
+    fold(recover, ::identity)
+  }
+
+  public suspend fun <R2> handleErrorWith(recover: suspend (R) -> Effect<R2, A>): Effect<R2, A> = effect {
+    fold({ recover(it).bind() }, ::identity)
+  }
+
+  public suspend fun <B> redeem(recover: suspend (R) -> B, transform: suspend (A) -> B): Effect<Nothing, B> =
+    effect {
+      fold(recover, transform)
+    }
+
+  public suspend fun <R2, B> redeemWith(
+    recover: suspend (R) -> Effect<R2, B>,
+    transform: suspend (A) -> Effect<R2, B>
+  ): Effect<R2, B> = effect { fold(recover, transform).bind() }
+}
+
+/**
+ * **AVOID USING THIS TYPE, it's meant for low-level cancellation code** When in need in low-level
+ * code, you can use this type to differentiate between a foreign [CancellationException] and the
+ * one from [Effect].
+ */
+public sealed class ShiftCancellationException : CancellationException("Shifted Continuation")
+
+//  Holds `R` and `suspend (R) -> B`, the exception that wins the race, will get to execute
+// `recover`.
+@PublishedApi
+internal class Internal(val token: Token, val shifted: Any?, val recover: suspend (Any?) -> Any?) :
+  ShiftCancellationException() {
+  override fun toString(): String = "ShiftCancellationException($message)"
+}
+
+// Class that represents a unique token by hash comparison
+@PublishedApi
+internal class Token {
```

Comment:
Can we re-use the existing `Token` that is already in Arrow?

Reply:
the only difference is that the one in arrow.generics is not transforming the hashCode with an unsigned Integer. If that's fine, I'll use it. Otherwise I could also change the one in arrow.generics?

Reply:
Ah, it's in `arrow-continuations` 👍 
That module will be deprecated, and dissapear over time so lets keep this one instead.

It's not necessary to call `toUInt()` first and so we can improve this code with `override fun toString(): String = "Token(${hashCode().toString(16)})"`. It was needed in an older Kotlin version IIRC.